### PR TITLE
[DRWTSN32] Initial rough implementation

### PR DESCRIPTION
**Pull request by learn-more**
## Purpose

This aims to add a rough sketch of how drwtsn32 could look like.
DrWtsn32 is a crash reporter.

## TODO

- [x] Figure out why the stacktrace does not work
- [x] Output to a file (on the desktop?) instead of stdout.
- [x] Figure out how to get the correct exception, instead of the attach breakpoint.
- [x] Notify the user about the stacktrace dumped on the desktop.
- [ ] Disassemble the offending function (maybe move the kdb disassembler into a library?)

### CODE REVIEW

**Comment by HBelusca** [comment]
\o/
You should also ask @encodedpr what the status of his crash-reporter is.

File: base/applications/drwtsn32/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
+PROJECT(drwtsn32)
+
+set_cpp(WITH_RUNTIME WITH_EXCEPTIONS WITH_STL)
+
+
```
Comment (HBelusca) [inline base/applications/drwtsn32/CMakeLists.txt]: Extra newline

File: base/applications/drwtsn32/drwtsn32.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
+#include "precomp.h"
```
Comment (HBelusca) [inline base/applications/drwtsn32/drwtsn32.cpp]: Missing code source copyright header.

File: base/applications/drwtsn32/drwtsn32.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+
+bool UpdateFromEvent(DEBUG_EVENT& evt, DumpData& data)
+{
+    switch(evt.dwDebugEventCode)
+    {
+    case CREATE_PROCESS_DEBUG_EVENT:
+    {
+        data.ProcessPath.resize(MAX_PATH);
+        DWORD len = GetModuleFileNameExA(evt.u.CreateProcessInfo.hProcess, NULL, &data.ProcessPath[0], data.ProcessPath.size());
```
Comment (HBelusca) [inline base/applications/drwtsn32/drwtsn32.cpp]: What about using UNICODE everywhere?
But then the STL stuff should be like std::wstring & co.
Unless you want to use ATL?

Comment (learn-more) [reply base/applications/drwtsn32/drwtsn32.cpp]: Eventually I want to print this to a file instead of stdout, and then UNICODE is not really convenient.

**Comment by encodedpr** [comment]
Status = STATUS_NOTHING

**Comment by yagoulas** [comment]
Here is the code of the logger I'm working on based on your code. I want it to be able to log a backtrace for the exceptions it may encounter so in the end the two will share a lot of code:

https://github.com/yagoulas/reactos/blob/roslogger/base/applications/roslogger/roslogger.cpp

**Comment by Mouvedia** [comment]
Please don't add an unique identifier.
cf https://www.schneier.com/blog/archives/2017/08/nsa_collects_ms.html

**Comment by HBelusca** [comment]
@Mouvedia , you are completely out of topic. That spying thingie is when reports are sent onto the internet to MS. Here the PR is about just a mere post-mortem debugger.

**Comment by Mouvedia** [comment]
@HBelusca if you are planning to send the crash reports to sometime in the future it will be relevant. That's a logical evolution.

**Comment by ThFabba** [comment]
It's a valid concern, although not for the current implementation.
However the "unique identifier" that article talks about is the IP address. It's, uhm... pretty hard to send things on the internet without including that.
Also, end-to-end encryption mitigates this problem, as the article also mentions.

**Comment by Mouvedia** [comment]
> However the "unique identifier" that article talks about is the IP address.

I was talking about `HKEY_LOCAL_MACHINE\Software\Microsoft\Cryptography\MachineGuid`.

**Comment by gigaherz** [comment]
What does the MachineGuid registry key have to do with this? This is a drwatson PR, implementing a debugging tool. That key is used exclusively by the cryptographic components.

**Comment by learn-more** [comment]
Please keep your tin foil hat BS out of this PR, thanks.

**Comment by learn-more** [comment]
The disassembly of the offending function will be added later, I want to go ahead and get this in.
Please review.

(After reviewing, all commits will be squashed to one)

File: base/applications/drwtsn32/drwtsn32.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                    if (data.Event)
+                    {
+                        SetEvent(data.Event);
+                        CloseHandle(data.Event);
+                        data.Event = NULL;
+                    }
+                    return true;
+                }
+                break;
+            case 0x406d1388:
```
Comment (HeisSpiter) [inline base/applications/drwtsn32/drwtsn32.cpp]: Magic value

Comment (learn-more) [reply base/applications/drwtsn32/drwtsn32.cpp]: Fair enough :)

**Pull request merged**